File: face/geomajas-face-dojo/client/src/main/resources/geomajas/gfx/painter/VectorLayerPainter.js
/*
 * This is part of Geomajas, a GIS framework, http://www.geomajas.org/.
 *
 * Copyright 2008-2011 Geosparc nv, http://www.geosparc.com/, Belgium.
 *
 * The program is available in open source according to the GNU Affero
 * General Public License. All contributions in this program are covered
 * by the Geomajas Contributors License Agreement. For full licensing
 * details, see LICENSE.txt in the project root.
 */

dojo.provide("geomajas.gfx.painter.VectorLayerPainter");
dojo.require("geomajas.gfx.Painter");

dojo.declare("VectorLayerPainter", Painter, {

	/**
	 * @class Painter implementation for layers. This painter always paints
	 * in worldspace!
	 * @author Jan De Moerloose & Pieter De Graef
	 *
	 * @constructor
	 * @extends Painter
	 */
	constructor : function (mapView) {
		this.mapView = mapView;
		this.labelStyle = new FontStyle("#FF0000", "10pt", "Courier New, Arial", "normal", "normal");
	},

	/**
	 * The actual painting function. It applies known layer styles, and also
	 * check the layer's visibility status. 
	 * @param layer A Paintable object.
	 * @param graphics A GraphicsContext object, responsible for actual drawing.
	 */
	paint : function (layer, graphics) {
		log.info("VectorLayerPainter.paint : painting layer "+layer.layerId + ", current scale = "+this.mapView.getCurrentScale());
        // Always redraw the layer style, because it can change anytime:
		graphics.drawGroup ({
			style:     layer.getDefaultStyle(),
			id:        layer.getId()
		});
		graphics.drawGroup ({id: layer.getId() + ".features"}); // create the group...
		graphics.drawGroup ({id: layer.getId() + ".selection"}); // create the group...
		graphics.drawGroup ({id: layer.getId() + ".labels"}); // create the group...

		// Draw symbol types, these can change anytime as well:
		var styles = layer.getStyles();
		for (var i=0; i<styles.count; i++) {
			var style = styles.item(i);
			graphics.drawShapeType({ style: style.getStyle(), id: style.styleId	});
		}

		// Check layer visibility:
		if (layer.checkVisibility(this.mapView.getCurrentScale())) {
			graphics.unhide (layer.getId());
		} else {
			graphics.hide (layer.getId());
		}

		// Check label visibility:
		if (layer.isLabeled()) {
			if (layer.getLabelFontStyle() != null) {
				var style = this.labelStyle.clone();
				style.setFillColor (layer.getLabelFontStyle().getFillColor());
				graphics.drawGroup ({id:layer.getId() + ".labels", style:style});
			} else {
				graphics.drawGroup ({id:layer.getId() + ".labels", style:this.labelStyle});
			}
			graphics.unhide (layer.getId() + ".labels");
		} else {
			graphics.hide (layer.getId() + ".labels");
		}
	},

	/**
	 * Deletes the layer. Actualy it only deletes the children. This overriding
	 * of the basic delete function is necessary to keep the map order.
	 */
	deleteShape : function (/*VectorLayer*/layer, /*GraphicsContext*/graphics) {
		log.error ("Deleting layer "+layer.getId());
		graphics.deleteShape(layer.getId(), true);
	},

	// Getters and setters:

	setLabelStyle : function (labelStyle) {
		this.labelStyle = labelStyle;;
	},
	
	getLabelStyle : function () {
		return this.labelStyle;
	}
});

File: client/geomajas-client-smartgwt/client/src/main/java/org/geomajas/smartgwt/client/gfx/painter/VectorLayerPainter.java
/*
 * This is part of Geomajas, a GIS framework, http://www.geomajas.org/.
 *
 * Copyright 2008-2013 Geosparc nv, http://www.geosparc.com/, Belgium.
 *
 * The program is available in open source according to the GNU Affero
 * General Public License. All contributions in this program are covered
 * by the Geomajas Contributors License Agreement. For full licensing
 * details, see LICENSE.txt in the project root.
 */

package org.geomajas.smartgwt.client.gfx.painter;

import org.geomajas.configuration.FeatureStyleInfo;
import org.geomajas.configuration.FontStyleInfo;
import org.geomajas.smartgwt.client.gfx.MapContext;
import org.geomajas.smartgwt.client.gfx.Paintable;
import org.geomajas.smartgwt.client.gfx.Painter;
import org.geomajas.smartgwt.client.gfx.style.FontStyle;
import org.geomajas.smartgwt.client.gfx.style.ShapeStyle;
import org.geomajas.smartgwt.client.map.layer.VectorLayer;
import org.geomajas.smartgwt.client.widget.MapWidget;
import org.geomajas.layer.LayerType;

/**
 * Painter for a VectorLayer object. Prepares the necessary groups for features, selected features and labels. Also
 * initiates shape-types, in case the vector layer is a point layer. On every draw, this painter will also check the
 * labeled and visible flags, and act accordingly.
 * 
 * @author Pieter De Graef
 */
public class VectorLayerPainter implements Painter {
	
	private MapWidget mapWidget;

	public VectorLayerPainter(MapWidget mapWidget) {
		this.mapWidget = mapWidget;
	}

	public String getPaintableClassName() {
		return VectorLayer.class.getName();
	}

	/**
	 * The actual painting function. Draws the groups.
	 * 
	 * @param paintable
	 *            A {@link VectorLayer} object.
	 * @param group
	 *            The group where the object resides in (optional).
	 * @param context
	 *            A MapContext object, responsible for actual drawing.
	 */
	public void paint(Paintable paintable, Object group, MapContext context) {
		VectorLayer layer = (VectorLayer) paintable;

		// Create the needed groups in the correct order:
		context.getVectorContext().drawGroup(mapWidget.getGroup(MapWidget.RenderGroup.VECTOR), layer);
		// layer.getDefaultStyle???
		context.getVectorContext().drawGroup(layer, layer.getFeatureGroup());
		context.getVectorContext().drawGroup(layer, layer.getSelectionGroup());
		FontStyle labelStyle = getLabelFontstyle(layer);
		context.getVectorContext().drawGroup(layer, layer.getLabelGroup(), labelStyle);

		// Create the needed groups in the correct order:
		context.getRasterContext().drawGroup(mapWidget.getGroup(MapWidget.RenderGroup.RASTER), layer);
		// layer.getDefaultStyle???
		context.getRasterContext().drawGroup(layer, layer.getFeatureGroup());
		context.getRasterContext().drawGroup(layer, layer.getLabelGroup());

		// Draw symbol types, as these can change any time:
		LayerType layerType = layer.getLayerInfo().getLayerType();
		if (LayerType.POINT == layerType || LayerType.MULTIPOINT == layerType || LayerType.GEOMETRY == layerType) {
			for (FeatureStyleInfo style : layer.getLayerInfo().getNamedStyleInfo().getFeatureStyles()) {
				context.getVectorContext().drawSymbolDefinition(null, style.getStyleId(), style.getSymbol(),
						new ShapeStyle(style), null);
			}
		}

		// Check layer visibility:
		if (layer.isShowing()) {
			context.getVectorContext().unhide(layer);
			context.getRasterContext().unhide(layer);
		} else {
			context.getVectorContext().hide(layer);
			context.getRasterContext().hide(layer);
		}

		// Check label visibility:
		if (layer.isLabelsShowing()) {
			context.getVectorContext().unhide(layer.getLabelGroup());
			context.getRasterContext().unhide(layer.getLabelGroup());
		} else {
			context.getVectorContext().hide(layer.getLabelGroup());
			context.getRasterContext().hide(layer.getLabelGroup());
		}
	}

	private FontStyle getLabelFontstyle(VectorLayer layer) {
		FontStyleInfo info = layer.getLayerInfo().getNamedStyleInfo().getLabelStyle().getFontStyle();
		return new FontStyle(info);
	}

	/**
	 * Delete a {@link Paintable} object from the given {@link MapContext}. It the object does not exist, nothing
	 * will be done.
	 * 
	 * @param paintable
	 *            The object to be painted.
	 * @param group
	 *            The group where the object resides in (optional).
	 * @param context
	 *            The context to paint on.
	 */
	public void deleteShape(Paintable paintable, Object group, MapContext context) {
		context.getVectorContext().deleteGroup(paintable);
		context.getRasterContext().deleteGroup(paintable);
	}

	// Getters and setters:

}
